cuda/matrix/ell_kernels.cu

```diff
@@ -151,17 +305,25 @@ void advanced_spmv(std::shared_ptr<const CudaExecutor> exec,
                    const matrix::Dense<ValueType> *beta,
                    matrix::Dense<ValueType> *c)
 {
-    const dim3 block_size(default_block_size, 1, 1);
-    const dim3 grid_size(ceildiv(a->get_size()[0], block_size.x),
-                         b->get_size()[1], 1);
-
-    abstract_spmv<<<grid_size, block_size, 0, 0>>>(
-        a->get_size()[0], as_cuda_type(alpha->get_const_values()),
-        as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
-        a->get_stride(), a->get_num_stored_elements_per_row(),
-        as_cuda_type(b->get_const_values()), b->get_stride(),
-        as_cuda_type(beta->get_const_values()), as_cuda_type(c->get_values()),
-        c->get_stride());
+    const auto data = compute_subwarp_size_and_atomicity(exec, a);
+    const int subwarp_size = std::get<0>(data);
+    const int atomic = std::get<1>(data);
+    const int nwarps_per_row = std::get<2>(data);
+
+    /**
+     * info is the parameter for selecting the cuda kernel.
+     * for info == 0, it uses the kernel by 32 threads with atomic operation
+     * for other value, it uses the kernel without atomic_add
+     */
+    const int info = (!atomic) * subwarp_size;
+    if (atomic) {
+        dense::scale(exec, beta, c);
+    }
+    select_abstract_spmv(
+        compiled_kernels(),
+        [&](int compiled_info) { return info == compiled_info; },
```

Comment:
Nit: I would only put `info` to the capture of the lambda, so it is clear where it comes from:
```suggestion
        [&info](int compiled_info) { return info == compiled_info; },
```